internal/app/orchestrator/orchestrator.go

```diff
@@ -95,16 +114,186 @@ func New(ctx context.Context,
-func (c *orchestrator) Fetch(context.Context, discovery.DiscoveryRequest) (*gcp.Response, error) {
+func (o *orchestrator) Fetch(context.Context, discovery.DiscoveryRequest) (*gcp.Response, error) {
 	return nil, fmt.Errorf("Not implemented")
 }
 
-func (c *orchestrator) onCacheEvicted(key string, resource cache.Resource) {
-	// TODO implement.
+// watchResponse is intended to be called in a goroutine, to receive incoming
+// responses and fan out to downstream clients.
+func (o *orchestrator) watchUpstream(
+	ctx context.Context,
+	aggregatedKey string,
+	responseChannel <-chan *discovery.DiscoveryResponse,
+	done <-chan bool,
+	shutdown func(),
+) {
+	for {
+		select {
+		case x, more := <-responseChannel:
+			if !more {
+				// A problem occurred fetching the response upstream, log retry.
+				// TODO implement retry/back-off logic on error scenario.
+				// https://github.com/envoyproxy/xds-relay/issues/68
+				o.logger.With("key", aggregatedKey).Error(ctx, "upstream error")
+				return
+			}
+			// Cache the response.
+			_, err := o.cache.SetResponse(aggregatedKey, *x)
+			if err != nil {
+				// TODO if set fails, we may need to retry upstream as well.
+				// Currently the fallback is to rely on a future response, but
+				// that probably isn't ideal.
+				// https://github.com/envoyproxy/xds-relay/issues/70
+				//
+				// If we fail to cache the new response, log and return the old one.
+				o.logger.With("err", err).With("key", aggregatedKey).
+					Error(ctx, "Failed to cache the response")
+			}
+
+			// Get downstream watches and fan out.
+			// We retrieve from cache rather than directly fanning out the
+			// newly received response because the cache does additional
+			// resource serialization.
+			cached, err := o.cache.Fetch(aggregatedKey)
+			if err != nil {
+				o.logger.With("err", err).With("key", aggregatedKey).Error(ctx, "cache fetch failed")
+				// Can't do anything because we don't know who the watches
+				// are. Drop the response.
+			} else {
+				if cached == nil || cached.Resp == nil {
+					// If cache is empty, there is nothing to fan out.
+					// Error. Sanity check. Shouldn't ever reach this since we
+					// just set the response, but it's a rare scenario that can
+					// happen if the cache TTL is set very short.
+					o.logger.With("key", aggregatedKey).Error(ctx, "attempted to fan out with no cached response")
+				} else {
+					// Goldenpath.
+					o.fanout(cached.Resp, cached.Requests)
+				}
+			}
+		case <-done:
+			// Exit when signaled that the stream has closed.
+			shutdown()
+			return
+		}
+	}
+}
+
+// fanout pushes the response to the response channels of all open downstream
+// watches in parallel.
+func (o *orchestrator) fanout(resp *cache.Response, watchers map[*gcp.Request]bool) {
+	for watch := range watchers {
+		go func(watch *gcp.Request) {
+			if channel, ok := o.downstreamResponseMap.get(watch); ok {
+				channel <- convertToGcpResponse(resp, *watch)
```

Comment:
need a default case?

Reply:
Is your proposal to remove it from the cache if we can't find a corresponding channel for the watch? This shouldn't happen if we're cleaning up properly but I can add the sanity check.

Reply:
the channel could be blocked for many reasons. if its blocked, we will need a default case, else the whole flow will be stuck on slowest response.

Reply:
since you are using gororutines, it wont be a problem, but imagine one slow channel. There will be multiple goroutines backed up to ssend response. We wont be able to guarentee which one goes first. If teh goroutines are rremoved, then default case will be necessary.

Reply:
I put a [comment below](https://github.com/envoyproxy/xds-relay/pull/53#discussion_r417697540) regarding the routines but I'll still add a default case. Maybe to timeout and log error if we're still not able to push to it after 10s?

Reply:
sounds good. Could use the `util.DoWithTimeout` for the purpose.